Eigen 3.4 vs VectorXd ValueProducer

Sub-task of #15142.

When we test with Eigen 3.4 on macOS, the compiler yells at us.

I don't understand what's happening yet, but here's the breadcrumb:
```
Eigen/src/Core/IndexedView.h:146:14: error: field type 'drake::systems::ContextBase' is an abstract class

while substituting deduced template arguments into function template 'ValueProducer' [with SomeOutput = drake::systems::LeafSystem<double> *]
```

-- https://drake-jenkins.csail.mit.edu/job/mac-catalina-clang-bazel-experimental-everything-debug/3/consoleText

The repro is just to pull #15142 locally _on macOS_ and try `bazel build //multibody/plant`.  I don't think Debug vs Release vs any other config settings matter, I think any kind of Apple Clangs are complaining here.

~Best guess, it's trying to copy something that it should be referencing instead?~

_Edit: New guess, seems like it's selecting the wrong `ValueProducer` overload.  We're trying to use overload (1b) with a `VectorXd` as a model value, but it's trying to do something else, probably (5b)._

--- a/systems/framework/value_producer.h
+++ b/systems/framework/value_producer.h
@@ -354,11 +354,12 @@ class ValueProducer final {
 
   /** Overload (5b). Refer to the
   @ref ValueProducer_constructors "Constructor overloads" for details. */
-  template <typename SomeOutput>
+  template <typename SomeOutput,
+      typename = std::enable_if_t<!std::is_convertible_v<
+          SomeOutput, AllocateCallback>>>
   ValueProducer(
-      const SomeOutput& model_value, CalcCallback calc,
-      std::enable_if_t<!std::is_convertible_v<
-          SomeOutput, AllocateCallback>>* = 0)
+      const SomeOutput& model_value,
+      CalcCallback calc)
       : ValueProducer(make_allocate_mode_b(model_value),
                       std::move(calc)) {}
 

